onnxruntime/contrib_ops/cuda/bert/group_query_attention.cc

```diff
@@ -0,0 +1,180 @@
+                             onnxruntime::flash::is_supported(device_prop,
+                                                              parameters.head_size,
+                                                              parameters.num_heads,
+                                                              parameters.kv_num_heads);
+  // Allocate buffers
+  size_t softmax_lse_bytes = 0;
+  size_t softmax_lse_accum_bytes = 0;
+  size_t out_accum_bytes = 0;
+  size_t seqlens_k_bytes = 0;
+  if (use_flash_attention) {
+    // Flash Attention kv-caching operates in place on a buffer, therefore past and present kv must be the same
+    assert(present_key == nullptr || present_key == past_key);
+    assert(present_value == nullptr || present_value == past_value);
+
+    softmax_lse_bytes = onnxruntime::flash::get_softmax_lse_size(parameters.sequence_length, parameters.batch_size, parameters.num_heads);
+    // split kv buffers
+    parameters.num_splits = onnxruntime::flash::num_splits_heuristic(
+        parameters.batch_size, parameters.sequence_length, parameters.kv_sequence_length, parameters.num_heads,
+        parameters.head_size, device_prop.multiProcessorCount, 128, false,
+        device_prop.major == 8 && device_prop.minor > 0);
+    if (parameters.num_splits > 1) {
+      // softmax_lse_accum buffer
+      softmax_lse_accum_bytes = onnxruntime::flash::get_softmax_lse_accum_size(
+          parameters.num_splits, parameters.batch_size, parameters.num_heads, parameters.sequence_length);
+      // out_accum buffer
+      auto round_multiple = [](int x, int m) { return (x + m - 1) / m * m; };
+      const int head_size_rounded = round_multiple(parameters.head_size, 32);
+      out_accum_bytes = onnxruntime::flash::get_out_accum_size(
+          parameters.num_splits, parameters.batch_size, parameters.num_heads, parameters.sequence_length, head_size_rounded);
+    }
+    // seqlens_k buffer
+    if (past_key != nullptr) {
+      seqlens_k_bytes = sizeof(int) * parameters.batch_size;
+    }
+  }
+  auto softmax_lse_buffer = GetScratchBuffer<void>(softmax_lse_bytes, context->GetComputeStream());
+  auto softmax_lse_accum_buffer = GetScratchBuffer<void>(softmax_lse_accum_bytes, context->GetComputeStream());
+  auto out_accum_buffer = GetScratchBuffer<void>(out_accum_bytes, context->GetComputeStream());
+  auto seqlens_k_buffer = GetScratchBuffer<void>(seqlens_k_bytes, context->GetComputeStream());
+#else
+  constexpr bool use_flash_attention = false;
+  auto softmax_lse_buffer = GetScratchBuffer<void>(0, context->GetComputeStream());        // nullptr
+  auto softmax_lse_accum_buffer = GetScratchBuffer<void>(0, context->GetComputeStream());  // nullptr
+  auto out_accum_buffer = GetScratchBuffer<void>(0, context->GetComputeStream());          // nullptr
+  auto seqlens_k_buffer = GetScratchBuffer<void>(0, context->GetComputeStream());          // nullptr
+#endif
+
+  // only kernel implemented for gqa right now
+  assert(use_flash_attention);
```

Comment:
Return an error status when flash attention is not applicable. It is because assert does not work in release build.